repose-aggregator/src/docs/asciidoc/filters/keystone-v2.adoc

```diff
@@ -1,52 +1,413 @@
 = Keystone v2 Filter
 
-include::../_includes/in-progress.adoc[]
+Provides a mechanism for authenticating incoming requests with the OpenStack Keystone v2 Identity service.
 
 == General filter information
-* *Name:* ???
-* *Default Configuration:* ???.cfg.xml
-* *Released:* v?.?.?.?
+* *Name:* keystone-v2
+* *Default Configuration:* keystone-v2.cfg.xml
+* *Released:* v7.1.5.0
 * *Bundle:* repose-filter-bundle
 * link:../schemas/keystone-v2.xsd[Schema]
 
 == Prerequisites & Postconditions
 === Required headers
-???.
+* `X-Auth-Token` - is a required header that is used by this filter to define the operating parameters of the HTTP transaction.
+If an incoming HTTP request is missing the `X-Auth-Token` header, then a response status code of _Unauthorized_ (`401`) is returned.
 
 === Required preceding filters
-???.
+This filter has no dependencies on other filters.
+
+However, it is a good practice to prevent spoofing of identities by putting the <<header-normalization.adoc#, Header Normalization filter>> before any authentication and/or authorization filters so that it can remove any headers that would be populated by them.
 
 === Request headers created
-???.
+The following headers are created using the information returned from the authentication service:
+
+* `X-Authorization` - Informs origin service that user has been authenticated. (ex. "Proxy User")
+* `X-Identity-Status` - Indicates if identity has been confirmed. (ex. "Confirmed", "Indeterminate")
+** Only if the `delegating` element is defined.
+* `X-User-Name` - Identifies user name. (ex. "jjenkins")
+* `X-User-ID` - Identifies user ID. (ex. "12345")
+* `X-Authenticated-By` - Identifies the method(s) by which the request was authenticated.  (ex. "password")
+* `X-Roles` - Identifies roles. (ex. "admin", "user")
+** This header is only populated if the `set-roles-in-header` attribute is `true` and there are roles returned from the authentication service.
+** If the `validate-tenant` element is present, then *Tenanted Mode* is active and this effects the contents of this header as outlined below:
+*** IF *NOT IN* Tenanted Mode, THEN all roles in the user's token will be forwarded.
+*** IF *IN* Tenanted Mode AND `pre-authorized`, THEN all roles in the user's token will be forwarded.
+**** In order for a user to be `pre-authorized` based on role, the role in question must either not have an associated tenant, or the associated tenant must match the tenant extracted from the URI.
```

Comment:
Yes, it is the case that roles associated with a tenant must match the tenant from the URI to be used for the `pre-authorized` check. If they are not associated with a tenant, then of course they do not need to match it as they are basically _global_ roles for that user.

> https://github.com/rackerlabs/repose/blob/927e229ae3b727f00c12c6ecd4ae0251ce1b6b48/repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala#L318-L327

And this is in line with the intent of the `pre-authorized` roles in that, if I have a role for tenant 123, then I should be pre-authorized based on that role when accessing tenant 456.